### PR TITLE
ensure pods roll when configmap is updated, better support for gitops

### LINKED ISSUE #14331: Add a configmap checksum to automatically roll daemonset
## Proposal / RFE

**Is your feature request related to a problem?**
I’m trying to rollout cilium via gitops tools such as argocd/fleet/etc. currently when just the configmap is updated by altering helm values I need to manually roll all the pods for the settings to take effect which leaves clusters in unknown states if automation is managing things.

**Describe the solution you'd like**
https://helm.sh/docs/howto/charts_tips_and_tricks/#automatically-roll-deployments

Simply add a checksum annotation to the daemonset pods and it will roll everything if a update to the configmap occurs.

Thanks!

### PR DESCRIPTION
**Pull request by travisghansen**
Please ensure your pull request adheres to the following guidelines:

- [X] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [X] All code is covered by unit and/or runtime tests where feasible.
- [X] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [X] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [X] Provide a title or release-note blurb suitable for the release notes.
- [X] Thanks for contributing!

Fixes: #14331 

```release-note
Helm Chart: add checksum to configmap to ensure pods roll when configmap is updated
```

Introduces the following new values:

```
rollOutCiliumPods: false

hubble:
  ui:
    rollOutPods: false
  relay:
    rollOutPods: false

operator:
  rollOutPods: false
```

### CODE REVIEW

**Comment by aanm** [comment]
Out of curiosity, isn't that guaranteed by helm? I guess not given the creation of this PR 😄

**Comment by travisghansen** [comment]
@aanm it's not actually and is a pretty common issue.

- https://helm.sh/docs/howto/charts_tips_and_tricks/#automatically-roll-deployments

**Comment by brb** [comment]
What happens if a user updates the ConfigMap via `kubectl edit cm cilium-config`, and then calls `helm update`? Will the checksum mechanism detect the change? Or does it rely on the presence of the ConfigMap file?

**Comment by travisghansen** [comment]
Editing manually would not roll pods. Helm wouldn’t care what’s in k8s, only what it renders at runtime so the hash would only change if something in chart changed to content of the configmap in some way.

Does that answer the question?

**Review by kaworu — CHANGES_REQUESTED** [review]
Can we have the same logic for Hubble Relay in `cilium/templates/hubble-relay-deployment.yaml` please?

File: install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml [hunk 1]
```diff
@@ -39,6 +39,8 @@ spec:
         prometheus.io/port: "{{ .Values.prometheus.port }}"
         prometheus.io/scrape: "true"
 {{- end }}
+        # ensure pods roll when configmap updates
+        checksum/cilim-configmap: {{ include (print $.Template.BasePath "/cilium-configmap.yaml") . | sha256sum }}
```
Comment (aanm) [inline install/kubernetes/cilium/templates/cilium-agent-daemonset.yaml]: s/`checksum/cilim-configmap:`/`cilium.io/configmap-checksum`

File: install/kubernetes/cilium/templates/cilium-operator-deployment.yaml [hunk 2]
```diff
@@ -27,6 +27,8 @@ spec:
   template:
     metadata:
       annotations:
+        # ensure pods roll when configmap updates
+        checksum/cilim-configmap: {{ include (print $.Template.BasePath "/cilium-configmap.yaml") . | sha256sum }}
```
Comment (aanm) [inline install/kubernetes/cilium/templates/cilium-operator-deployment.yaml]: s/`checksum/cilim-configmap:`/`cilium.io/configmap-checksum`

**Review by aanm — CHANGES_REQUESTED** [review]
Just a small nit in the annotation name. Also, you need to run `make -C install/kubernetes` to re-generate the `quick-install.yaml`

**Comment by ti-mo** [comment]
Adding my 2c: this behaviour might surprise some adminstrators that are used to being able to modify the configmap without having all agents roll automatically. For example, you might want to roll some agents as a canary to make sure they can start properly with the new config.

Auto-restart imposes certain risks, as some aspects of L7 packet processing (e.g. resolving FQDN policies) depend on the agent being up to function correctly. IMO this feature should be disabled by default. For me, this feature would be most useful when making frequent changes to the configmap on new clusters/environments where I don't need to worry about causing impact due to misconfigs yet.

We've discussed the option of live config reloading internally, which would definitely be the safer option. Not sure how far off we are from being able to deliver this. @aanm

**Comment by aanm** [comment]
> Adding my 2c: this behaviour might surprise some adminstrators that are used to being able to modify the configmap without having all agents roll automatically. For example, you might want to roll some agents as a canary to make sure they can start properly with the new config.
> 
> Auto-restart imposes certain risks, as some aspects of L7 packet processing (e.g. resolving FQDN policies) depend on the agent being up to function correctly. IMO this feature should be disabled by default. For me, this feature would be most useful when making frequent changes to the configmap on new clusters/environments where I don't need to worry about causing impact due to misconfigs yet.
> 
> We've discussed the option of live config reloading internally, which would definitely be the safer option. Not sure how far off we are from being able to deliver this. @aanm

Good points, perhaps would be better to have it as a opt-in behind a helm flag and have some documentation about this? Which place would be better? :thinking:

**Comment by travisghansen** [comment]
All great points! I considered making it optional but wasn’t sure where would be appropriate place to put the flag in values. If someone has a suggestion I’ll add it in there and get everything else cleaned up.

Regarding hubble relay I’ll check it out, I simply grepped the project for usages so may have missed things. Any other places that use the configmap that I missed?

**Comment by aanm** [comment]
> All great points! I considered making it optional but wasn’t sure where would be appropriate place to put the flag in values. If someone has a suggestion I’ll add it in there and get everything else cleaned up.
> 
I would suggest these locations?

```diff
diff --git a/install/kubernetes/cilium/values.yaml b/install/kubernetes/cilium/values.yaml
index ea817b31f..945a880af 100644
--- a/install/kubernetes/cilium/values.yaml
+++ b/install/kubernetes/cilium/values.yaml
@@ -64,6 +64,8 @@ agent: true
 # -- Agent container name.
 name: cilium
 
+rolloutCiliumPods: true
+
 # -- Agent container image.
 image:
   repository: quay.io/cilium/cilium
@@ -468,6 +470,8 @@ hubble:
   # -- Enable Hubble (true by default).
   enabled: true
 
+  rollOutPods: true
+
   # -- Buffer size of the channel Hubble uses to receive monitor events. If this
   # value is not set, the queue size is set to the default monitor queue size.
   # eventQueueSize: ""
@@ -1075,6 +1079,8 @@ operator:
   # -- Enable the cilium-operator component (required).
   enabled: true
 
+  rollOutPods: true
+
   # -- cilium-operator image.
   image:
     repository: quay.io/cilium/operator
```


> Regarding hubble relay I’ll check it out, I simply grepped the project for usages so may have missed things. Any other places that use the configmap that I missed?

@kaworu ?

**Comment by kaworu** [comment]
@aanm
> I would suggest these locations?

About the second part of your suggestion (i.e. `hubble.rollOutPods`): as Hubble is part of the agent deployment, I fail to see what it would be used for.

Relay has its own ConfigMap, so looking at the current state of the diff I would suggest something like:

```diff
diff --git a/install/kubernetes/cilium/templates/hubble-relay-deployment.yaml b/install/kubernetes/cilium/templates/hubble-relay-deployment.yaml
index 8edad06b7..693d69c00 100644
--- a/install/kubernetes/cilium/templates/hubble-relay-deployment.yaml
+++ b/install/kubernetes/cilium/templates/hubble-relay-deployment.yaml
@@ -18,8 +18,9 @@ spec:
 {{- end }}
   template:
     metadata:
-{{- with .Values.hubble.relay.podAnnotations }}
       annotations:
+        checksum/cilim-configmap: {{ include (print $.Template.BasePath "/hubble-relay-configmap.yaml") . | sha256sum }}
+{{- with .Values.hubble.relay.podAnnotations }}
         {{ toYaml . | trim | indent 8 }}
 {{- end }}
       labels:

```

Of course to be adapted with a knob like `hubble.relay.rollOutPods` to conditionally roll-out, and @aanm's suggestion about the name.

**Comment by travisghansen** [comment]
OK, I'll have a look and see if I can get everything cleaned up. Thanks!

**Comment by travisghansen** [comment]
I've never used the preflight features, should the checksum be added to `cilium/templates/cilium-preflight-daemonset.yaml`?

Also, technically, mounted `secrets` should do the same thing...any opinions on adding those checksums as well?

**Comment by kaworu** [comment]
> I've never used the preflight features, should the checksum be added to `cilium/templates/cilium-preflight-daemonset.yaml`?
> 
> Also, technically, mounted `secrets` should do the same thing...any opinions on adding those checksums as well?

About the hubble and relay secrets: we should not add checksums for them. They are the TLS keypairs and both hubble and relay have support for hot-reloading them.

**Comment by aanm** [comment]
> I've never used the preflight features, should the checksum be added to `cilium/templates/cilium-preflight-daemonset.yaml`?
> 
> Also, technically, mounted `secrets` should do the same thing...any opinions on adding those checksums as well?

Regarding the preflight, I don't think it's necessary. The preflight is used before performing an upgrade where no configmap changes are involved.

**Comment by travisghansen** [comment]
Does cilium agent and hubble relay already automatically update as a result of changes to `hubble-ca-configmap.yaml`?

**Comment by kaworu** [comment]
> Does cilium agent and hubble relay already automatically update as a result of changes to hubble-ca-configmap.yaml?

@travisghansen correct, both the cilium agent and relay watch the file mounted by the `hubble-ca-cert` ConfigMap to handle hot-reloading on update.

From the cilium agent (i.e. hubble observer server) side, the CA is used to authenticate mTLS clients such as Relay (`option.Config.HubbleTLSClientCAFiles`):
https://github.com/cilium/cilium/blob/c6b4a5e137b1078667b8b63a293f1132337bf619/daemon/cmd/hubble.go#L188-L193

Relay use it to authenticate the hubble observer server (`keyTLSHubbleServerCAFiles`):
https://github.com/cilium/cilium/blob/c6b4a5e137b1078667b8b63a293f1132337bf619/hubble-relay/cmd/serve/serve.go#L158-L163

**Comment by travisghansen** [comment]
OK cool, if that's watched in the application then I won't bother adding checksums for the configmaps...updated tweaks will be pushed in the next couple days then.

**Comment by travisghansen** [comment]
@ti-mo having hot reload is more ideal in the sense that pods/processes do not restart. However I would make note that k8s rolling can be ‘safer’ in certain contexts. For example, if the config is corrupted in some shape or form hot reload gives the operator no ability to rollout in a controlled fashion (one mess up and the whole cluster could be a complete mess). On the flip side, k8s rolling the pods will automatically stop the rollout if the pods start hitting crashloops etc.

There is no ‘silver bullet’ here but we could add options for turning on/off the hot reload in code and keep the on/off switch on the checksums in the chart I guess getting us closer to silver bullet.

Maybe others have more suggestions that could create a more complete approach adding to the above.

**Review by kaworu — CHANGES_REQUESTED** [review]
Patch LGTM, but I think we should quote those hashes just to be safe, i.e. replace `| sha256sum` by `| sha256sum | quote`.

**Comment by travisghansen** [comment]
OK, I've added `quote` to all of the checksums.

**Review by kaworu — APPROVED** [review]
Thanks @travisghansen patch LGTM now!

File: install/kubernetes/experimental-install.yaml [hunk 3]
```diff
@@ -1079,6 +1079,7 @@ spec:
     type: RollingUpdate
   template:
     metadata:
+      annotations:
```
Comment (ti-mo) [inline install/kubernetes/experimental-install.yaml:1082]: Just curious, why are these `annotations` keys explicitly declared here?

Comment (travisghansen) [reply install/kubernetes/experimental-install.yaml:1082]: It’s from changing the structure of the chart logic slightly and then running the `make` commands. Previously some of the annotation flow logic was a bit strange and in some cases only facilitated annotations being added explicitly by the user in values.yaml...it’s a byproduct of ‘fixing’ that.

**Review by qmonnet — APPROVED** [review]
Looks good to me, thank you.

File: install/kubernetes/cilium/README.md [hunk 4]
```diff
@@ -184,6 +184,7 @@ contributors across the globe, there is almost always someone available to help.
 | hubble.relay.replicas | int | `1` | Number of replicas run for the hubble-relay deployment. |
 | hubble.relay.resources | object | `{}` | Specifies the resources for the hubble-relay pods |
 | hubble.relay.retryTimeout | string | `nil` | Backoff duration to retry connecting to the local hubble instance in case of failure (e.g. "30s"). |
+| hubble.relay.rollOutPods | bool | `false` | Roll Hubble Relay pods automatically when configmap is updated. |
```
Comment (kkourt) [inline install/kubernetes/cilium/README.md]:
```suggestion
| hubble.relay.rollOutPods | bool | `false` | Rollout Hubble Relay pods automatically when configmap is updated. |
```

File: install/kubernetes/cilium/README.md [hunk 5]
```diff
@@ -213,6 +214,7 @@ contributors across the globe, there is almost always someone available to help.
 | hubble.ui.proxy.image | object | `{"pullPolicy":"Always","repository":"docker.io/envoyproxy/envoy","tag":"v1.14.5"}` | Hubble-ui ingress proxy image. |
 | hubble.ui.proxy.resources | object | `{}` |  |
 | hubble.ui.replicas | int | `1` |  |
+| hubble.ui.rollOutPods | bool | `false` | Roll Hubble-ui pods automatically when configmap is updated. |
```
Comment (kkourt) [inline install/kubernetes/cilium/README.md]:
```suggestion
| hubble.ui.rollOutPods | bool | `false` | Rollout Hubble-ui pods automatically when configmap is updated. |
```

File: install/kubernetes/cilium/README.md [hunk 6]
```diff
@@ -282,6 +284,7 @@ contributors across the globe, there is almost always someone available to help.
 | operator.prometheus.serviceMonitor.enabled | bool | `false` | Enable service monitors. This requires the prometheus CRDs to be available (see https://github.com/prometheus-operator/prometheus-operator/blob/master/example/prometheus-operator-crd/monitoring.coreos.com_servicemonitors.yaml) |
 | operator.replicas | int | `2` | Number of replicas to run for the cilium-operator deployment |
 | operator.resources | object | `{}` | cilium-operator resource limits & requests ref: https://kubernetes.io/docs/user-guide/compute-resources/ |
+| operator.rollOutPods | bool | `false` | Roll cilium-operator pods automatically when configmap is updated. |
```
Comment (kkourt) [inline install/kubernetes/cilium/README.md]:
```suggestion
| operator.rollOutPods | bool | `false` | Rollout cilium-operator pods automatically when configmap is updated. |
```

File: install/kubernetes/cilium/README.md [hunk 7]
```diff
@@ -317,6 +320,7 @@ contributors across the globe, there is almost always someone available to help.
 | remoteNodeIdentity | bool | `true` | Enable use of the remote node identity. ref: https://docs.cilium.io/en/v1.7/install/upgrade/#configmap-remote-node-identity |
 | resourceQuotas | object | `{"cilium":{"hard":{"pods":"10k"}},"enabled":false,"operator":{"hard":{"pods":"15"}}}` | Enable resource quotas for priority classes used in the cluster. |
 | resources | object | `{}` | Agent resource limits & requests ref: https://kubernetes.io/docs/user-guide/compute-resources/ |
+| rollOutCiliumPods | bool | `false` | Roll cilium agent pods automatically when configmap is updated. |
```
Comment (kkourt) [inline install/kubernetes/cilium/README.md]:
```suggestion
| rollOutCiliumPods | bool | `false` | Rollout cilium agent pods automatically when configmap is updated. |
```

File: install/kubernetes/cilium/values.yaml [hunk 8]
```diff
@@ -64,6 +64,9 @@ agent: true
 # -- Agent container name.
 name: cilium
 
+# -- Roll cilium agent pods automatically when configmap is updated.
```
Comment (kkourt) [inline install/kubernetes/cilium/values.yaml]:
```suggestion
# -- Rollout cilium agent pods automatically when configmap is updated.
```

File: install/kubernetes/cilium/values.yaml [hunk 9]
```diff
@@ -563,6 +566,9 @@ hubble:
     # -- Enable Hubble Relay (requires hubble.enabled=true)
     enabled: false
 
+    # -- Roll Hubble Relay pods automatically when configmap is updated.
```
Comment (kkourt) [inline install/kubernetes/cilium/values.yaml]:
```suggestion
    # -- Rollout Hubble Relay pods automatically when configmap is updated.
```

File: install/kubernetes/cilium/values.yaml [hunk 10]
```diff
@@ -640,6 +646,9 @@ hubble:
   ui:
     enabled: false
 
+    # -- Roll Hubble-ui pods automatically when configmap is updated.
```
Comment (kkourt) [inline install/kubernetes/cilium/values.yaml]:
```suggestion
    # -- Rollout Hubble-ui pods automatically when configmap is updated.
```

File: install/kubernetes/cilium/values.yaml [hunk 11]
```diff
@@ -1075,6 +1084,9 @@ operator:
   # -- Enable the cilium-operator component (required).
   enabled: true
 
+  # -- Roll cilium-operator pods automatically when configmap is updated.
```
Comment (kkourt) [inline install/kubernetes/cilium/values.yaml]:
```suggestion
  # -- Rollout cilium-operator pods automatically when configmap is updated.
```

**Review by kkourt — APPROVED** [review]
Looks good to me, thanks!

Just a minor nit for your consideration, "rollout" seems a bit better than just "roll" to me, but the meaning should be obvious to people in either case, so feel free to ignore the suggestions.

**Comment by travisghansen** [comment]
Sounds great. I’ll update the comments in code and regen so that future rebuilds will have proper values and repush the branch.

**Review by aanm — APPROVED** [review]
Nothing else besides what others already mentioned.

**Comment by travisghansen** [comment]
OK, I've updated the branch. Note I used `Roll out` instead of `Rollout` to align with the camel-casing of the parameters in `values.yaml`. If we don't want that let me know.

**Comment by ti-mo** [comment]
> `Roll out` instead of `Rollout`

Thanks, this is the way it should be. :) Was meaning to add that.

**Comment by aanm** [comment]
test-me-please

**Comment by travisghansen** [comment]
Anything I need to do to get this over the hill?

**Comment by gandro** [comment]
> Anything I need to do to get this over the hill?

I believe this only needs to pass CI, people from the relevant sub-teams have already approved. The previous run hit some unrelated flakes, I will restart them.

Runtime hit #14463: https://jenkins.cilium.io/job/Cilium-PR-Runtime-4.9/3333/testReport/junit/(root)/Suite-runtime/RuntimeFQDNPolicies_Enforces_ToFQDNs_policy/
4.19 had a provisioning failure:. https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-Kernel/4471/

**Comment by gandro** [comment]
retest-runtime

**Comment by gandro** [comment]
retest-4.19

**Comment by Rolinh** [comment]
I think enough people reviewed this PR and the CI is green. Marking as ready to merge.

**Pull request merged**